### PR TITLE
Add ccache-report target to standard Makefile

### PR DESCRIPTION
**Pull request by gezalore**
Using the standard model Makefile, when in addition to an explicit
target, the target 'ccache-report' is also given, a summary of ccache
hits/misses during this invocation of 'make' will be prited at the end
of the build.

This is quite handy working on ccache debugging, would you like to mainline it?

### CODE REVIEW

File: include/verilated.mk.in [hunk 1]
```diff
@@ -35,6 +35,7 @@ CFG_LDLIBS_THREADS = @CFG_LDLIBS_THREADS@
 
 VERILATOR_COVERAGE = $(PERL) $(VERILATOR_ROOT)/bin/verilator_coverage
 VERILATOR_INCLUDER = $(PERL) $(VERILATOR_ROOT)/bin/verilator_includer
+VERILATOR_CCREPORT = $(VERILATOR_ROOT)/bin/verilator_ccache_report
```
Comment (wsnyder) [inline include/verilated.mk.in:38]:
```suggestion
VERILATOR_CCACHE_REPORT = $(PYTHON3) $(VERILATOR_ROOT)/bin/verilator_ccache_report
```

File: test_regress/driver.pl [hunk 2]
```diff
@@ -2304,6 +2304,10 @@ sub cfg_with_threaded {
     return 1;  # C++11 now always required
 }
 
+sub cfg_with_ccache {
+    return `grep "OBJCACHE \?= ccache" "../include/verilated.mk"` ne "";
```
Comment (wsnyder) [inline test_regress/driver.pl:2308]:
```suggestion
    return `grep "OBJCACHE \?= ccache" "$ENV{VERILATOR_ROOT}/include/verilated.mk"` ne "";
```

File: bin/verilator_ccache_report [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+######################################################################
+
+import argparse
+import collections
+import re
+
+parser = argparse.ArgumentParser(
+    allow_abbrev=False,
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    description="""Report ccache behaviour of verilated model build.""",
```
Comment (wsnyder) [inline bin/verilator_ccache_report:12]:
```suggestion
    description="""Report ccache behavior of a Verilated model build.""",
```

File: bin/verilator_ccache_report [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    line = line.strip()
+    match = re.match(r'.*Object file: (.*)$', line)
+    if match:
+        obj = match.group(1)
+    match = re.match(r'.*Result: (.*)$', line)
+    if match:
+        results[obj] = match.group(1)
+
+args.o.write("#" * 80 + "\n")
+args.o.write("ccache report:\n")
```
Comment (wsnyder) [inline bin/verilator_ccache_report:38]:
```suggestion
args.o.write("ccache report (from verilator_ccache_report) :\n")
```

File: include/verilated.mk.in [hunk 5]
```diff
@@ -254,6 +255,37 @@ endif
...
+######################################################################
+### ccache report
+
+ifneq ($(findstring ccache-report,$(MAKECMDGOALS)),)
+  ifneq ($(OBJCACHE),ccache)
+    $(error ccache-report requires OBJCACHE to equal 'ccache')
+  endif
+  VK_OTHER_GOALS := $(strip $(subst ccache-report,,$(MAKECMDGOALS)))
+  ifeq ($(VK_OTHER_GOALS),)
+    $(error ccache-reprot must be used with at least one other explicit target)
```
Comment (wsnyder) [inline include/verilated.mk.in:267]:
```suggestion
    $(error ccache-report must be used with at least one other explicit target)
```

**Review by wsnyder — COMMENTED** [review]
- [x] If you haven't please be sure to run "make py-lint" and clean up anything resulting (there's no regression for this at present).
- [ ] Add information on what this does and how to use it under docs/guide/ please.  (Unless you think we'd never have users run this... but I think we might?)
- [ ] Add bin/verilator_ccache_report to t_flag_help.pl

**Comment by wsnyder** [comment]
- [ ] Fix test failure (obviously ;)

**Comment by gezalore** [comment]
All done I think. Wasn't sure which part of the manual would be best to add this to.

**Review by wsnyder — COMMENTED** [review]
Thanks, please also add a new file/section similar to exe_verilator_profcfuncs.rst (and add references to it in what you just wrote).

Then good to merge.

**Comment by gezalore** [comment]
> Thanks, please also add a new file/section similar to exe_verilator_profcfuncs.rst (and add references to it in what you just wrote).

Are you sure? verailator_ccache_report is more like verilator_includer and like that script, it lives in `share/verilator/bin` in an installation (not in `bin`), so it's not even on the path by default. I don't think the user ever wants to invoke this themselves, it's really us just a noddy script because I didn't want to write the logfile parsing in make.

**Comment by wsnyder** [comment]
You are installing it.  But I guess you're right it's like verilator_includer, so probably ok not to have a rst file for it.

>( cd ${srcdir}/bin ; $(INSTALL_PROGRAM) verilator_ccache_report $(DESTDIR)$(pkgdatadir)/bin/verilator_ccache_report )

**Review by wsnyder — COMMENTED** [review]
Feel free to squash merge when happy.

**Pull request merged**